### PR TITLE
8315127: CDSMapTest fails with incorrect number of oop references

### PR DESCRIPTION
**Pull request by iklam**
This is a bug in the test case. The test parses patterns like this in the map file

```
0x00000000ffe00000: @@ Object (0xffe00000) java.lang.String
 - klass: 'java/lang/String' 0x0000000800010220
 - fields (3 words):
 - private 'hash' 'I' @12  0 (0x00000000)
 - private final 'coder' 'B' @16  0 (0x00)
 - private 'hashIsZero' 'Z' @17  true (0x01)
 - injected 'flags' 'B' @18  1 (0x01)
 - private final 'value' '[B' @20 0x00000000ffe00018 (0xffe00018) [B length: 0
0x00000000ffe00018: @@ Object (0xffe00018) [B length: 0
 - klass: {type array byte} 0x00000008000024d8
```

Before this fix, the test assets that there are at least 10000 oop references (such as the `value` field in the above String). However, some JDK builds may have fewer archived heap objects, resulting in less than 10000 references (the graal build in the bug report has only 8286 references).

After this fix, we check that the number of oop references is not fewer than the number of strings, as we know each string contains one oop field. This is sufficient to check that the map file contains expected output, without using a hard-coded number.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315127](https://bugs.openjdk.org/browse/JDK-8315127): CDSMapTest fails with incorrect number of oop references (**Bug** - P4)

### Reviewers
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15527/head:pull/15527` \
`$ git checkout pull/15527`

Update a local copy of the PR: \
`$ git checkout pull/15527` \
`$ git pull https://git.openjdk.org/jdk.git pull/15527/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15527`

View PR using the GUI difftool: \
`$ git pr show -t 15527`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15527.diff">https://git.openjdk.org/jdk/pull/15527.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15527#issuecomment-1701944892)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/cds/CDSMapReader.java [hunk 1]
```diff
@@ -282,16 +309,29 @@ public static int validate(MapFile mapFile) {
                     //     To see how an invalidate pointer may be found, change oop in the
                     //     following line to oop+1
                     mustContain(mapFile.oopToObject, field, oop, false);
-                    count ++;
+                    count1 ++;
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/CDSMapReader.java:312]: Could `oop` be 0? If so, `oop` could be overcounted.

Pre-existing: extra import at line 27
`import java.io.RandomAccessFile;`

File: test/hotspot/jtreg/runtime/cds/CDSMapTest.java [hunk 2]
```diff
@@ -76,11 +76,6 @@ static void dump(ArrayList<String> args, String... more) throws Exception {
         CDSTestUtils.createArchiveAndCheck(opts);
 
         CDSMapReader.MapFile mapFile = CDSMapReader.read(mapName);
-        int oopFieldCount = CDSMapReader.validate(mapFile);
-        if (mapFile.heapObjectCount() > 0 && oopFieldCount < 10000) {
-            // heapObjectCount() may be zero if the selected GC doesn't support heap object archiving.
-            throw new RuntimeException("CDS map file seems incorrect: " + mapFile.heapObjectCount() +
-                                       " objects but only " + oopFieldCount + " oop field references");
-        }
+        CDSMapReader.validate(mapFile);
```
Comment (calvinccheung) [inline test/hotspot/jtreg/runtime/cds/CDSMapTest.java:79]: Pre-existing: extra import at lines 36 and 37
```
import java.io.FileInputStream;
import java.io.IOException;
```

**Review by calvinccheung — COMMENTED** [review]
I have one question and spotted two minor issues.

File: test/hotspot/jtreg/runtime/cds/CDSMapReader.java [hunk 3]
```diff
@@ -254,7 +279,8 @@ public static MapFile read(String fileName) {
             throw new RuntimeException(t);
         } finally {
             System.out.println("Parsed " + lineCount + " lines in " + fileName);
-            System.out.println("Found "  + mapFile.heapObjectCount() + " heap objects");
+            System.out.println("Found "  + mapFile.heapObjectCount() + " heap objects ("
```
Comment (turbanoff) [inline test/hotspot/jtreg/runtime/cds/CDSMapReader.java:282]:
```suggestion
            System.out.println("Found " + mapFile.heapObjectCount() + " heap objects ("
```

File: test/hotspot/jtreg/runtime/cds/CDSMapReader.java (hunk 1 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/CDSMapReader.java:312]: I added checks for `oop != 0`, so now the test is counting only the non-null references.

File: test/hotspot/jtreg/runtime/cds/CDSMapTest.java (hunk 2 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/CDSMapTest.java:79]: Fixed.

File: test/hotspot/jtreg/runtime/cds/CDSMapReader.java (hunk 3 above)
Comment (iklam) [reply test/hotspot/jtreg/runtime/cds/CDSMapReader.java:282]: Fixed.

**Review by calvinccheung — APPROVED** [review]
Thanks for the update. It looks good.

**Comment by iklam** [comment]
Thanks @calvinccheung and @turbanoff for the review and comments.
/integrate

**Pull request closed without merging**